### PR TITLE
Add new pass for math to rocdl.

### PR DESCRIPTION
**Pull request by jsjodin**
There are math operations that are not yet being converted, but it works for most operations. The changes to the flang codegen enables the pass for AMDGCN and disables the MathToLLVM and MathToLibm passes.

### CODE REVIEW

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 1]
```diff
@@ -3559,6 +3560,14 @@ class FIRToLLVMLowering
     // as passes here.
     mlir::OpPassManager mathConvertionPM("builtin.module");
 
+    bool isAMDGCN = fir::getTargetTriple(mod).isAMDGCN();
+    // If compiling for AMD target some math operations must be lowered to ocml
```
Comment (agozillon) [inline flang/lib/Optimizer/CodeGen/CodeGen.cpp:3564]: Nit: would it be possible to get the full name for ocml in the comments? Maybe a weird request, just unfamiliar with the library so would be handy to have in the comment! if not that's not a big deal.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 2]
```diff
@@ -3567,6 +3576,7 @@ class FIRToLLVMLowering
     mathConvertionPM.addPass(
         mlir::createConvertMathToFuncs(mathToFuncsOptions));
     mathConvertionPM.addPass(mlir::createConvertComplexToStandardPass());
+
```
Comment (agozillon) [inline flang/lib/Optimizer/CodeGen/CodeGen.cpp:3579]: Nit: Unnecessary space addition

**Review by agozillon — APPROVED** [review]
The PR LGTM :-) feel free to wait on a secondary reviewer approval or not, I'll leave it up to you!

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp (hunk 1 above)
Comment (jsjodin) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:3564]: I think ocml stands for OpenCL Math Library, so I'm not sure that will help or confuse more. I should perhaps use "AMD GPU math library" or something like that.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 3]
```diff
@@ -3559,6 +3560,14 @@ class FIRToLLVMLowering
     // as passes here.
     mlir::OpPassManager mathConvertionPM("builtin.module");
 
+    bool isAMDGCN = fir::getTargetTriple(mod).isAMDGCN();
+    // If compiling for AMD target some math operations must be lowered to ocml
+    // library calls, the rest can be converted to LLVM intrinsics, which is
+    // handled in the mathToLLVM conversion. The lowering to libm calls is not
+    // needed since all math operations are handled this way.
+    if (isAMDGCN) {
```
Comment (skatrak) [inline flang/lib/Optimizer/CodeGen/CodeGen.cpp:3568]: Nit: Braces

**Review by skatrak — APPROVED** [review]
Thank you Jan, LGTM as well. It'd be good to add a Lit test for this as well, to make sure it doesn't inadvertently break later on.

**Pull request merged**